chore(sdk): Use codecov's CircleCI orb instead of custom nox session

timoffex opened the pull request:
Description
-----------
Orb documented at https://circleci.com/developer/orbs/orb/codecov/codecov

Testing
-------
Wait to see codecov results on this PR.

kptkin reviewed (approved):
We were using args to allow for other arguments but we didn't pass anything so we can revisit it when we need

timoffex commented:
It looks like their upload command doesn't work on Mac, but it seems like a simple bug. The command

```sh
echo "${CODECOV_PUBLIC_PGP_KEY}" | \
  gpg --no-default-keyring --keyring trustedkeys.kbx --import
```

seems to not create the "trustedkeys.kbx" file, causing a subsequent command to fail.

Mac failure: https://app.circleci.com/pipelines/github/wandb/wandb/30862/workflows/665304f8-a40e-4ca5-96e3-1df0b793069d/jobs/1009775/steps

Linux success: https://app.circleci.com/pipelines/github/wandb/wandb/30862/workflows/665304f8-a40e-4ca5-96e3-1df0b793069d/jobs/1009770

I'll investigate it more tomorrow.

kptkin commented:
> It looks like their upload command doesn't work on Mac, but it seems like a simple bug. The command
> 
> ```shell
> echo "${CODECOV_PUBLIC_PGP_KEY}" | \
>   gpg --no-default-keyring --keyring trustedkeys.kbx --import
> ```
> 
> seems to not create the "trustedkeys.kbx" file, causing a subsequent command to fail.
> 
> Mac failure: https://app.circleci.com/pipelines/github/wandb/wandb/30862/workflows/665304f8-a40e-4ca5-96e3-1df0b793069d/jobs/1009775/steps
> 
> Linux success: https://app.circleci.com/pipelines/github/wandb/wandb/30862/workflows/665304f8-a40e-4ca5-96e3-1df0b793069d/jobs/1009770
> 
> I'll investigate it more tomorrow.

is it the `validate` flag, i don't think we did it before, don't know if it is required...

kptkin commented:
> It looks like their upload command doesn't work on Mac, but it seems like a simple bug. The command
> 
> ```shell
> echo "${CODECOV_PUBLIC_PGP_KEY}" | \
>   gpg --no-default-keyring --keyring trustedkeys.kbx --import
> ```
> 
> seems to not create the "trustedkeys.kbx" file, causing a subsequent command to fail.
> 
> Mac failure: https://app.circleci.com/pipelines/github/wandb/wandb/30862/workflows/665304f8-a40e-4ca5-96e3-1df0b793069d/jobs/1009775/steps
> 
> Linux success: https://app.circleci.com/pipelines/github/wandb/wandb/30862/workflows/665304f8-a40e-4ca5-96e3-1df0b793069d/jobs/1009770
> 
> I'll investigate it more tomorrow.

i think we can skip this step if we  set the `validate` argument to false in the codecov orb. Not sure if we want to do it (for reference we didn't verify the uploader when we were downloading it manually in nox)

kptkin commented on .circleci/config.yml:
```diff
@@ -514,12 +513,9 @@ jobs:
                   not:
-                    equal: ["", << parameters.codecov_args >>]
+                    equal: ["", << parameters.codecov_flags >>]
                 steps:
-                  - run:
-                      name: Upload coverage to codecov
-                      command: |
-                        nox -s codecov -- << parameters.codecov_args >>
+                  - codecov/upload: {flags: <<parameters.codecov_flags>>}
```
I saw that the codecov orb has a `when` argument so in theory we could use it, up to you ofc

timoffex commented:
@kptkin - Could we remove "ci/circleci: system-linux-py37" from the required workflows? I realized that it cannot pass because I removed core-py37 in a previous PR (with core-py38 being the min). The code coverage step throws an error now when it sees that there was no code coverage info emitted.

kptkin commented:
> @kptkin - Could we remove "ci/circleci: system-linux-py37" from the required workflows? I realized that it cannot pass because I removed core-py37 in a previous PR (with core-py38 being the min). The code coverage step throws an error now when it sees that there was no code coverage info emitted.

i guess but it means that we say that we support py37 but we don't actually test it anywhere

kptkin commented:
> > @kptkin - Could we remove "ci/circleci: system-linux-py37" from the required workflows? I realized that it cannot pass because I removed core-py37 in a previous PR (with core-py38 being the min). The code coverage step throws an error now when it sees that there was no code coverage info emitted.
> 
> i guess but it means that we say that we support py37 but we don't actually test it anywhere

I think you removed it because reports needed 3.8, but we can put it back in tox and have these dependencies specified better, you can use requirements specifiers

timoffex commented:
> > > @kptkin - Could we remove "ci/circleci: system-linux-py37" from the required workflows? I realized that it cannot pass because I removed core-py37 in a previous PR (with core-py38 being the min). The code coverage step throws an error now when it sees that there was no code coverage info emitted.
> > 
> > 
> > i guess but it means that we say that we support py37 but we don't actually test it anywhere
> 
> I think you removed it because reports needed 3.8, but we can put it back in tox and have these dependencies specified better, you can use requirements specifiers (https://pip.pypa.io/en/stable/reference/requirement-specifiers/)

Ah good point. Added back the py37 system tests.

kptkin commented:
> > > > @kptkin - Could we remove "ci/circleci: system-linux-py37" from the required workflows? I realized that it cannot pass because I removed core-py37 in a previous PR (with core-py38 being the min). The code coverage step throws an error now when it sees that there was no code coverage info emitted.
> > > 
> > > 
> > > i guess but it means that we say that we support py37 but we don't actually test it anywhere
> > 
> > 
> > I think you removed it because reports needed 3.8, but we can put it back in tox and have these dependencies specified better, you can use requirements specifiers (https://pip.pypa.io/en/stable/reference/requirement-specifiers/)
> 
> Ah good point. Added back the py37 system tests.

our python support policy is: we keep the python version 1-1.5 years after EOF (https://devguide.python.org/versions/), so i think we can start thinking of removing 3.7 sometimes mid year.
